firebase_functions/functions/src/helpers.ts

```diff
@@ -163,11 +163,72 @@ const faucet = async receiveAddress => {
   );
 };
 
+
+const getBalance = async address => {
+  try {
+    if (!address) {
+      return 0;
+    }
+    const { provider } = await getSettings();
+    const { getBalances } = composeAPI({ provider });
+    const { balances } = await getBalances([address], 100);
+    return balances && balances.length > 0 ? balances[0] : 0;
+  } catch (error) {
+    console.error('getBalance error', error);
+    return 0;
+  }
+};
+
+
+const repairWallet = async (seed, keyIndex) => {
+  try {
+    const address = await generateAddress(seed, keyIndex)
+    const balance = await getBalance(address)
+  
+    if (balance === 0){
+    const iterable = [-2, -1, 0, 1, 2, 3, 4, -3, -4,-5,-6-7,5,6,7];
+
+    for (let value of iterable) {
+
+      if ((value < 0 && Math.abs(value) <= keyIndex) || value >= 0) {
```

Comment:
Given the current keyIndex of the Iota Wallet of 2685, the expression `Math.abs(value) <= keyIndex` with values up to 7 does'n really makes sense

Reply:
also, expression  `(value < 0 && true) || value >= 0)` will be always executed, since all possible cases are covered. 
Given that, you don't need the `if` statement

Reply:
This function is not only written for an IOTA wallet with a keyIndex > 2685.
If we have a wallet (like in the IMP) with a keyIndex of 1 and we are taking a value of -2 we would receive a keyIndex of -1, which will give an Error. 

This case is covered by the if loop:
-2 < 0  true 
Math.abs(-2) = 2 
2 <= 1 False 

value -2 >= 0 False

Reply:
but you are using the IOTA faucet wallet, same as for DMP, to fund IMP wallets. So its keyIndex is the same

Reply:
every time you access the `getIotaWallet`, its keyIndex will be 2685 or larger, as of now
```
  let { keyIndex, seed, defaultBalance } = await getIotaWallet();
  const newIotaWallet :any = await repairWallet(seed, keyIndex)
```

Reply:
Not just for the faucet, also for regular wallets of an user within the IMP

Reply:
Function `repairWallet` used in this file 2 times. In both cases it is called from a function that funds a wallet (DMP user or IMP user, does not matter) from the IOTA wallet. 
In both cases you need to repair the IOTA wallet to be able to use the faucet function. 
In both cases the keyIndex of the IOTA wallet is > 2000

Reply:
Also note, that IMP has its own firebase project. So if you are not using the IOTA faucet to fund a wallet, but just want to restore balance of the existing wallet, its repairWallet function should be placed these. It does not belong to DMP project

Reply:
Every user has a wallet within the IMP. If the wallet stays on pending, the wallet within the database also needs to be repaired. So my idea was to create the function in a way that it is reusable within different scenarios, not only for the funding process. But if this approach is that pointless for you, I can also remove this line.

If the function is approved, I can add it to the IMP. I think it doesn't really make sense to create two PRs with the same function?